app/controllers/financialStatement/ChargeDetailsController.scala

```diff
@@ -45,42 +48,49 @@ class ChargeDetailsController @Inject()(identify: IdentifierAction,
     with I18nSupport
     with NunjucksSupport {
 
-  def onPageLoad(identifier: String, startDate: LocalDate, chargeReference: String): Action[AnyContent] = identify.async {
+  def onPageLoad(identifier: String, startDate: LocalDate, index: String): Action[AnyContent] = identify.async {
     implicit request =>
       fsConnector.getPsaFS(request.psaId.id).flatMap {
         psaFS =>
           val filteredPsaFS = psaFS.filter(_.periodStartDate == startDate)
+          fiCacheConnector.fetch flatMap {
+            case Some(jsValue) =>
+              val chargeRefs: Seq[String] = (jsValue \ "chargeRefs").as[Seq[String]]
+
+              val commonJson = Json.obj(
+                "heading" -> heading(psaFS.filter(_.chargeReference == chargeRefs(index.toInt)).head.chargeType.toString),
+                "isOverdue" -> penaltiesService.isPaymentOverdue(psaFS.filter(_.chargeReference == chargeRefs(index.toInt)).head),
+                "period" -> msg"penalties.period".withArgs(startDate.format(dateFormatterStartDate),
+                  getQuarter(startDate).endDate.format(dateFormatterDMY)),
+                "chargeReference" -> psaFS.filter(_.chargeReference == chargeRefs(index.toInt)).head.chargeReference,
```

Comment:
I'd use headOption and redirect to session expired page if index out of range

Reply:
Done